src/libcore/future/future.rs

```diff
@@ -99,6 +99,18 @@ pub trait Future {
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>;
 }
 
+/// Conversion into a `Future`.
+#[unstable(feature = "into_future", issue = "0")]
+pub trait IntoFuture {
+    /// Which kind of future are we turning this into?
+    #[unstable(feature = "into_future", issue = "0")]
+    type Future: Future;
+
+    /// Creates a future from a value.
+    #[unstable(feature = "into_future", issue = "0")]
+    fn into_future(self) -> Self::Future;
+}
+
```

Comment:
[`std::iter::IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) has an associated `Item` type:

```rust
pub trait IntoIterator {
    type Item;
    type IntoIter: Iterator;
    fn into_iter(self) -> Self::IntoIter;
}
```

I'm thinking it would make sense for `IntoFuture` to be consistent with it and similarly define an `Output` type:

```rust
pub trait IntoFuture {
    type Output;
    type Future: Future<Output = Self::Output>;
    fn into_future(self) -> Self::Future;
}
```

It's not a big deal, but having consistency between similar traits generally seems desirable. This is also why we've chosen that direction for async-std's [`IntoFuture`](https://github.com/async-rs/async-std/pull/259) and [`IntoStream`](https://docs.rs/async-std/0.99.9/async_std/stream/trait.IntoStream.html) APIs.

Reply:
The reason for this is that it can help avoid double indirection in associated types, `<<T as IntoFuture>::Future as Future>::Output` is quite gnarly, and its very nice to be able to write `<T as IntoFuture>::Output` instead.

Reply:
I agree with this proposal.

Reply:
Done.